drivers/resolver/ad2s1210/ad2s1210.c

```diff
@@ -0,0 +1,361 @@
+#include "no_os_delay.h"
+#include "no_os_alloc.h"
+
+const char *ad2s1210_device_name[] = {
+	"AD2S1210",
+};
+
+static int32_t ad2s1210_set_mode_pins(struct ad2s1210_dev *dev,
+				      enum ad2s1210_mode mode);
+
+uint32_t ad2s1210_reg_write(struct ad2s1210_dev *dev, uint8_t addr, uint8_t val)
+{
+	uint8_t buf;
+	int ret;
+
+	if (addr < AD2S1210_REG_MIN)
+		return -EINVAL;
+
+	ad2s1210_set_mode_pins(dev, MODE_CONFIG);
+	buf = addr;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+	if (ret)
+		return ret;
+
+	buf = val;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+uint32_t ad2s1210_reg_read(struct ad2s1210_dev *dev, uint8_t addr, uint8_t *val)
+{
+	uint8_t buf;
+	int32_t ret;
+
+	if (addr < AD2S1210_REG_MIN)
+		return -EINVAL;
+
+	ad2s1210_set_mode_pins(dev, MODE_CONFIG);
+	buf = addr;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+	if (ret)
+		return ret;
+
+	buf = addr;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+	if (ret)
+		return ret;
+
+	*val = buf;
+	return 0;
+}
+
+static int32_t ad2s1210_set_mode_pins(struct ad2s1210_dev *dev,
+				      enum ad2s1210_mode mode)
+{
+	int32_t ret;
+
+	if (dev->mode == mode)
+		return 0;
+
+	ret = no_os_gpio_set_value(dev->gpio_a0, mode & NO_OS_BIT(0));
+	if (ret)
+		return ret;
+
+	ret = no_os_gpio_set_value(dev->gpio_a1, mode & NO_OS_BIT(1));
+	if (ret)
+		return ret;
+
+	dev->mode = mode;
+	return 0;
+}
+
+static int32_t ad2s1210_set_resolution(struct ad2s1210_dev *dev,
+				       uint8_t resolution)
+{
+	int32_t ret;
+	uint8_t control;
+
+	ret = ad2s1210_reg_read(dev, AD2S1210_REG_CONTROL, &control);
+	if (ret)
+		return ret;
+
+	control &= ~(AD2S1210_REG_CONTROL_RES_MASK);
+	switch (resolution) {
+	case 10:
+		control |= 0x0;
+		break;
+	case 12:
+		control |= 0x1;
+		break;
+	case 14:
+		control |= 0x2;
+		break;
+	case 16:
+		control |= 0x3;
+		break;
+	default:
+		pr_err("Invalid resolution: %d\n", resolution);
+		return -EINVAL;
+	}
+
+	ret = ad2s1210_reg_write(dev, AD2S1210_REG_CONTROL, control);
+	if (ret)
+		return ret;
+
+
+	if (!dev->have_resolution_pins)
+		return 0;
+
+	ret = no_os_gpio_set_value(dev->gpio_res0, control & NO_OS_BIT(0));
+	if (ret)
+		return ret;
+
+	ret = no_os_gpio_set_value(dev->gpio_res1, !!(control & NO_OS_BIT(1)));
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static int32_t ad2s1210_gpio_init(struct no_os_gpio_desc **desc,
+				  struct no_os_gpio_init_param *param,
+				  uint8_t value,
+				  bool is_optional)
+{
+	int32_t ret;
+
+	if (is_optional && !param->platform_ops)
+		return 0;
+
+	if (is_optional)
+		ret = no_os_gpio_get_optional(desc, param);
+	else
+		ret = no_os_gpio_get(desc, param);
+	if (ret)
+		return ret;
+
+	if (!desc)
+		return 0;
+
+	ret = no_os_gpio_direction_output(*desc, value);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static void ad2s1210_gpio_remove(struct ad2s1210_dev *dev)
+{
+	if (dev->gpio_a0)
+		no_os_gpio_remove(dev->gpio_a0);
```

Comment:
no error check handling. same for the rest of the gpio_remove functions

Reply:
ok will add.